Enhancements for macOS Support in Schnitzel Motor

runstop opened the pull request:
This commit introduces support for the macOS platform in the Schnitzel Motor engine. Noteworthy changes include:
- The build.sh script has been expanded to:
  * Identify macOS as a system
  * Configure platform-specific variables
  * Apply appropriate C flags for the macOS platform.
- macOS-specific source files, namely mac_platform.h and mac_platform.mm, have been added. These handle:
  * macOS-specific code tasks, such as window creation and updates.
  * macOS Window rendering uses Cocoa framework. running ./schnitzel from terminal launches Schnitzel Motor just as it does on other platform.
- The .gitignore file has been updated to ignore macOS-specific files and directories.
- A preprocessor directive has been added to main.cpp to include the relevant platform header file based on the operating system detected.
- The previous  declaration was not compatible with macOS. The  declaration not only resolves this issue but works as before for all supported platforms.

Testing:
- Tested changes by running build.sh without error on Windows, Linux and Mac.
- Successfully launched schnitzel on Windows, Linux and Mac.

Cakez77 committed:
Enhancements for macOS Support in Schnitzel Motor

This commit introduces support for the macOS platform in the Schnitzel Motor engine.
Noteworthy changes include:
- The build.sh script has been expanded to:
  * Identify macOS as a system
  * Configure platform-specific variables
  * Apply appropriate C flags for the macOS platform.
- macOS-specific source files, namely mac_platform.h and mac_platform.mm, have been added. These handle:
  * macOS-specific code tasks, such as window creation and updates.
  * macOS Window rendering uses Cocoa framework. running ./schnitzel from terminal launches Schnitzel Motor just as it does on other platform.
- The .gitignore file has been updated to ignore macOS-specific files and directories.
- A preprocessor directive has been added to main.cpp to include the relevant platform header file based on the operating system detected.
- The previous  declaration was not compatible with macOS. The  declaration not only resolves this issue but works as before for all supported platforms.

Testing:
- Tested changes by running build.sh without error on Windows, Linux and Mac.
- Successfully launched schnitzel on Windows, Linux and Mac.

Cakez77 committed:
Enhancements for macOS Support in Schnitzel Motor

This commit introduces support for the macOS platform in the Schnitzel Motor engine.
Noteworthy changes include:
- The build.sh script has been expanded to:
  * Identify macOS as a system
  * Configure platform-specific variables
  * Apply appropriate C flags for the macOS platform.
- macOS-specific source files, namely mac_platform.h and mac_platform.mm, have been added. These handle:
  * macOS-specific code tasks, such as window creation and updates.
  * macOS Window rendering uses Cocoa framework. running ./schnitzel from terminal launches Schnitzel Motor just as it does on other platform.
- The .gitignore file has been updated to ignore macOS-specific files and directories.
- A preprocessor directive has been added to main.cpp to include the relevant platform header file based on the operating system detected.
- The previous  declaration was not compatible with macOS. The  declaration not only resolves this issue but works as before for all supported platforms.

Testing:
- Tested changes by running build.sh without error on Windows, Linux and Mac.
- Successfully launched schnitzel on Windows, Linux and Mac.

Cakez77 committed:
Added logging and a game.cpp with an update loop

Cakez77 commented:
Waiting on update about what we talked about. Also thank you very much for doing this!

Cakez77 committed:
test

runstop committed:
Mac Platform Pull-Request changes:

Refactor build.sh and Main.cpp, Add mac_platform.cpp, and Update .gitignore

- Simplified  script to reduce complexity:
  *  variable refactored to a more simplified form
  *  variable updated to  for Linux
  * Added a custom clang call for Mac to ensure compiler understands bridge dependencies

- Added , to map C++ platform functions to equivalent Objective-C calls on macOS

- Refactored  to remove need for special defines for Mac platform functions

- Updated  to ignore object files

runstop committed:
- Logger changes to support Mac: Added Apple debug break to DEBUG_BREAK call
- Changed quit app keyboard shortcut to 'cmd-q'
- Mac build
  * Was not able to defer compiling  objective-c code, as the linker would fail to find the objc platform functions
  * Added a clean-up .o operation prior to build
    Declaring static bool  creates a variable with internal linkage. This visibility is limited to the specific .cpp file, preventing its use as a shared variable across multiple files.
    I was not able to replace 'extern bool running' back to static bool, as the variable would not properly communicate state changes to 'running' from the objective-c file to main.cpp.
  Fixed a bug in platform_update_window_objc where the function looped infinitely, preventing execution from passing up to the main loop.

runstop commented:
I've modified my pull request to improve simplicity and readability, based on yesterday's feedback from the stream.  

- Avoid creating a distinct platform.h file for each platform. A single platform.h file should be used across all platforms to prevent confusion.
- Refrain from introducing multiple files into the compiler to ensure shorter compile times. The objective is to compile a single main.cpp file, not various ones.
- The usage of the source_files variable is discouraged. It was perceived as more confusing than useful. It's recommended to revert the source main cpp to its initial position.
- The variable project files should be eliminated. The focus should be on a single output file.
- An extern bool running should be implemented and set to true in the MacOS platform code.
- The libs variable should be kept in its original form, unless MacOS demands a different setup.

I've tried implementing as many of the suggestions as I could. Where I fell short, it was not due to a lack of trying, but rather because I could not get the app to build for all platforms without making the concessions I ended up submitting in my latest push. I'm happy to continue working on these changes until we can get the Mac code up to par.

runstop commented on src/mac_platform.m:
```diff
@@ -34,25 +34,23 @@ bool platform_create_window_objc(int width, int height, char* title)
 
 void platform_update_window_objc()
 {
-    while (running) {
```
Fixed bug where window held program execution unnecessarily. Now execution gets back to main loop.

runstop commented on src/mac_platform.m:
```diff
@@ -2,7 +2,7 @@
 static NSWindow* window;
 
-bool running = true; // time to signal exit from main loop
+extern bool running;  // <- declare, but do not define
```
Needed to declare extern to properly communicate state back to main loop on Mac platform, but more importantly this appears to be the proper variable declaration for how 'running' is used across multiple files.

runstop commented on src/linux_platform.cpp:
- Changed so that window is created with flexible dimensions specified by the width and height arguments, instead of using fixed values.
- Added declaration of 'running' variable allowing it's state to be shared and accessed across multiple files within the program.

runstop commented on src/mac_platform.cpp:
Bridge Purpose: The bridge acts as an interface between C++ and Objective-C code, allowing them to communicate and interact with each other. It enables the use of Objective-C code within a C++ project or vice versa, facilitating interoperability between the two languages.

runstop commented on src/main.cpp:
```diff
@@ -1,29 +1,31 @@
 // #############################################################################
 //                           Platform Globals
 // #############################################################################
-static bool running = true;
```
Removing the static keyword from the running variable declaration allows 'running' to be shared and accessed from other C++ files, ensuring that changes made to the variable's state in one file can be observed in other files as well.

runstop commented on .gitignore:
```diff
@@ -5,10 +5,19 @@ project.4coder
 build
 .vscode
 .vs
-/schnitzel*
```
Just removed blank space, not schnitzel* entry as it might first appear

runstop commented on build.sh:
During my build tests, I found I needed to added the "src/mac_platform.m" to clang because of the extra layer of conversion that is need to bridge platform functions to the objective-c mappings. I also added a clean step to ensure old Mac_platform.o don't impact the current build. This is only on the Mac build, and does not impact of platform builds. 

Compiler flag warnings, I added,  -Wno-c++11-extensions and -Wno-deprecated-declarations, to suppress specific warning messages from clang. These flags are specifically for macOS but do not impact other platform other platform builds.

runstop commented on src/schnitzel_lib.h:
```diff
@@ -23,6 +23,8 @@
 #define DEBUG_BREAK() __debugbreak()
 #elif __linux__
 #define DEBUG_BREAK() __asm__ volatile ("int3")
+#elif __APPLE__
```
Added Mac version of debug break command

runstop commented on src/linux_platform.cpp:
```diff
@@ -1,6 +1,6 @@
 
 #include <X11/Xlib.h>
-
+extern bool running;  // <- declare, but do not define
```
Needed to declare extern to properly communicate state back to main loop on Mac platform, but more importantly this appears to be the proper variable declaration for how 'running' is used across multiple files.

Cakez77 reviewed (approved):
Thank you very much for your time and effort you put into this. Let's hope it works together now.
